src/levanter/compat/torch_serialization.py

```diff
@@ -366,22 +366,46 @@ def to_numpy_state_dict(model, prefix: Optional[str] = None) -> StateDict:
     This method is especially useful for saving models distributed across multiple hosts.
     """
 
-    def get_to_cpu(arr):
-        if not is_jax_array_like(arr):
-            return arr
-        elif isinstance(arr, np.ndarray):
-            return arr
-        elif arr.is_fully_addressable:
-            r = np.array(arr)
-            return r
-        else:
-            return np.array(jax.device_get(multihost_utils.process_allgather(arr, tiled=True)))
+    with jax.default_device(jax.local_devices(backend="cpu")[0]):
 
-    # need to make sure the model is on *this machine* and *this machine's CPU* before saving
-    model = jax.tree_map(lambda arr: get_to_cpu(arr), model)
-    # TODO: it's be nice if safetensors supported an iterator or something so we could do the allgather one at a time
-    state_dict = model.to_state_dict(prefix=prefix)
-    return state_dict
+        def get_to_cpu(arr):
+            if not is_jax_array_like(arr):
+                return arr
+            elif isinstance(arr, np.ndarray):
+                return arr
+            elif arr.is_fully_addressable:
+                r = np.array(arr)
+                return r
+            else:
+                # unfortunately, jax's allgather seems to replicate to every device rather than every host
+                # which doesn't work for ~7B parameter models on TPU (assuming we also have optimizer state)
+                # this approach limits us to <64B parameters, but that's good enough for now
+                # we're going to do something a bit fancy, where we shard make a (process, device) mesh,
+                # then look for some axis along which we can shard the array, and then we'll do an allgather
+                # via pjit. If we can't find one, we'll just fully replicate since it probably isn't that big.
+                # TODO: ensure that this mesh arranges devices correctly
```

Comment:
How do we do that?

Reply:
eh, i dunno. i'm pretty sure this is guaranteed by JAX